edge-hub/watchdog/linux/src/main.rs

```diff
@@ -0,0 +1,185 @@
+                    self.name, e
+                );
+                false
+            }
+        }
+    }
+
+    fn wait_for_exit(&mut self) {
+        let mut elapsed_secs = 0;
+        while elapsed_secs < PROCESS_SHUTDOWN_TOLERANCE_SECS && self.is_running() {
+            let poll_interval: Duration = Duration::new(PROCESS_POLL_INTERVAL_SECS, 0);
+            thread::sleep(poll_interval);
+            elapsed_secs += PROCESS_POLL_INTERVAL_SECS;
+        }
+    }
+}
+
+fn main() -> Result<(), Error> {
+    init_logging();
+    info!("Starting watchdog");
+
+    let should_shutdown = match register_shutdown_listener() {
+        Ok(should_shutdown) => should_shutdown,
+        Err(e) => {
+            error!(
+                "Failed to register sigterm listener. Shutting down. {:?}",
+                e
+            );
+            exit(1);
+        }
+    };
+
+    let has_broker_shutdown = ChildProcess::run_child_process(
+        "MQTT Broker".to_string(),
+        Command::new("/usr/local/bin/mqttd").stdout(Stdio::inherit()),
+        Arc::clone(&should_shutdown),
+    );
+
+    let has_edgehub_shutdown = ChildProcess::run_child_process(
+        "Edge Hub".to_string(),
+        Command::new("dotnet")
+            .arg("/app/Microsoft.Azure.Devices.Edge.Hub.Service.dll")
+            .stdout(Stdio::inherit()),
+        Arc::clone(&should_shutdown),
+    );
+
+    while !should_shutdown.load(Ordering::Relaxed) {
+        if has_broker_shutdown.load(Ordering::Relaxed) {
+            break;
+        }
+
+        if has_edgehub_shutdown.load(Ordering::Relaxed) {
+            break;
+        }
+
+        thread::sleep(Duration::from_secs(1));
+    }
+
+    should_shutdown.store(true, Ordering::Relaxed); // tell the threads to shut down their child process
+
+    // TODO: should we block on the thread exit? We don't have access to the threads intentionally though
+    let mut elapsed_secs = 0;
+    let buffered_wait_secs = PROCESS_SHUTDOWN_TOLERANCE_SECS + 5; // give buffer to allow child processes to be killed
+    while elapsed_secs < buffered_wait_secs
+        && (!has_broker_shutdown.load(Ordering::Relaxed)
+            || !has_edgehub_shutdown.load(Ordering::Relaxed))
+    {
+        let poll_interval: Duration = Duration::new(PROCESS_POLL_INTERVAL_SECS, 0);
+        thread::sleep(poll_interval);
+        elapsed_secs += PROCESS_POLL_INTERVAL_SECS;
+    }
+
+    info!("Stopped watchdog process");
+    Ok(())
+}
+
+fn init_logging() {
+    let subscriber = Subscriber::builder().with_max_level(Level::INFO).finish();
+    let _ = subscriber::set_global_default(subscriber);
+}
+
+// TODO: figure out how to log shutdown signal
+fn register_shutdown_listener() -> Result<Arc<AtomicBool>, Error> {
+    info!("Registering shutdown signal listener");
+    let should_shutdown = Arc::new(AtomicBool::new(false));
+    flag::register(SIGTERM, Arc::clone(&should_shutdown))?;
+    flag::register(SIGINT, Arc::clone(&should_shutdown))?;
```

Comment:
Using this signature we cannot log when the signal is received and of what type. I am going to look for a way to do this.